packages/test/test-end-to-end-tests/src/test/dataMigration.spec.ts

```diff
@@ -0,0 +1,425 @@
+			const detachedContext = createReadOnlyContext(context, "v2");
+			this._currentRuntimeFactory = this.v1ToV2;
+			return this.currentRuntimeFactory.preInitialize(detachedContext, existing);
+		}
+
+		return this.currentRuntimeFactory.preInitialize(context, existing);
+	}
+
+	public async instantiateFirstTime(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFirstTime(runtime);
+	}
+
+	public async instantiateFromExisting(runtime: ContainerRuntime): Promise<void> {
+		return this.currentRuntimeFactory.instantiateFromExisting(runtime);
+	}
+
+	public async hasInitialized(runtime: ContainerRuntime): Promise<void> {
+		await this.currentRuntimeFactory.hasInitialized(runtime);
+
+		if (this.clientDetailsType === summarizerClientType) {
+			this.oldRuntime = runtime;
+			const readonlyRuntime = await runtime.loadDetachedAndTransitionFn();
+			this.deferred.resolve({ summarizer: this.oldRuntime, readonlyRuntime });
+		}
+	}
+}
+
+class SharedCellToSharedMapFactory implements IChannelFactory {
+	public static readonly Type = "https://graph.microsoft.com/types/cell";
+
+	public get type(): string {
+		return SharedCellToSharedMapFactory.Type;
+	}
+
+	public get attributes(): IChannelAttributes {
+		return SharedMap.getFactory().attributes;
+	}
+
+	public async load(
+		runtime: IFluidDataStoreRuntime,
+		id: string,
+		services: IChannelServices,
+		attributes: IChannelAttributes,
+	): Promise<SharedMap> {
+		const cell = new SharedCell(id, runtime, attributes);
+		await cell.load(services);
+		const map = SharedMap.getFactory().create(runtime, id) as SharedMap;
+		const val = cell.get();
+		map.set(cellKey, val as string);
+		return map;
+	}
+
+	public create(document: IFluidDataStoreRuntime, id: string): SharedMap {
+		throw new Error("Shouldn't be making a shared cell");
+	}
+}
+
+async function logTree(tree: ISnapshotTree, storage: IDocumentStorageService) {
+	const newTree: any = {};
+	newTree.blobs = {};
+	newTree.trees = {};
+	for (const [blobName, blobId] of Object.entries(tree.blobs)) {
+		const content = await readAndParse(storage, blobId);
+		newTree.blobs[blobName] = content;
+	}
+	for (const [treeName, treeNode] of Object.entries(tree.trees)) {
+		newTree.trees[treeName] = await logTree(treeNode, storage);
+	}
+	// eslint-disable-next-line @typescript-eslint/no-unsafe-return
+	return newTree;
+}
+
+async function migrate(summarizerRuntime: ContainerRuntime, readonlyRuntime: ContainerRuntime) {
+	(readonlyRuntime as any).summarizerNode.startSummary(
+		readonlyRuntime.deltaManager.lastSequenceNumber,
+		new TelemetryNullLogger(),
+	);
+	const appSummary = await readonlyRuntime.summarize({ fullTree: true });
+	readonlyRuntime.disposeFn();
+	const neverCancel = new Deferred<SummarizerStopReason>();
+	const runningSummarizer = (summarizerRuntime as any).summarizer.runningSummarizer;
+	await runningSummarizer.summarizingLock;
+	await runningSummarizer.lockedSummaryAction(
```

Comment:
See comment above - I think this kind of locking should happen immediately in response to V2 acceptance op, such that summarizer can't produce one more summary with ref seq number of V2 acceptance op. That's fine as prototype - this is mostly invitation for discussion on how to organize code in the future to have proper correctness gurantees.

Reply:
Yeah, the locking should happen on receival of the op.